[Core] Switch Handlers

jsuarezruiz opened the pull request:
### Description of Change ###

Switch Handlers.

| MAUI API | Android | iOS | 
| ----|:--------:|:---:|
| IsToggled |  ✅  | ✅  | 
| TrackColor (OnColor) |  ✅  | ✅  | 
| ThumbColor| ✅  | ✅  |  

### Platforms Affected ### 

- Core
- iOS
- Android

### PR Checklist ###

- [x] Targets the correct branch
- [x] Include Handler tests
- [ ] Tests are passing (or failures are unrelated)

jsuarezruiz commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Handlers/Switch/SwitchHandler.cs:
```diff
@@ -0,0 +1,28 @@
+			[nameof(ISwitch.ThumbColor)] = MapThumbColor
+		};
+
+		public static void MapIsToggled(SwitchHandler handler, ISwitch view) => handler.TypedNativeView?.UpdateIsToggled(view);
```
We can have all the logic of the Handler in a single class. However, this will require quite a few compilation directives:
- Namespaces.
- CreateView.
- Teardown.
- GetDesiredSize.
- Specific native control events, etc.

This makes reading the code more complicated.
For that reason, I think a simple way to maintain everything is:
- Map all the properties in the common handler class.
- Use extension methods to do the necessary logic.
- Have the specific platform code in .Android, .iOS, etc. classes.

Have sense?

jsuarezruiz replied:
Reviewed and approved in a team meeting.

rmarinho commented on Xamarin.Forms.Platform.Android/AppCompat/SwitchRenderer.cs:
```diff
@@ -0,0 +1,172 @@
+using System;
```
why are you adding this file?

jsuarezruiz replied:
Merge error. Fixed. This file no longer exists.

rmarinho commented on src/Platform.Renderers/src/Xamarin.Forms.Platform.Android/Extensions/DrawableExtensions.cs:
```diff
@@ -72,36 +75,52 @@ public static void SetColorFilter(this ADrawable drawable, AColorFilter colorFil
 		{
-			if (Forms.Is29OrNewer)
-				drawable.SetColorFilter(new BlendModeColorFilter(color, GetFilterMode(mode)));
+#if __ANDROID_29__
```
this is not the same thing as `Forms.Is29OrNewer` is it?

rmarinho commented on src/Platform.Renderers/src/Xamarin.Forms.Platform.Android/Extensions/DrawableExtensions.cs:
```diff
@@ -1,23 +1,22 @@
+#endif
 
-namespace Xamarin.Forms.Platform.Android
+namespace Xamarin.Platform
```
If this is Xamarin.Platform shouldn't this move to Platform.Handlers folder =?

jsuarezruiz replied:
`DrawableExtensions.cs` have been moved to Platform.Handlers.

rmarinho commented on src/Platform.Renderers/src/Xamarin.Forms.Platform.Android/AppCompat/SwitchRenderer.cs:
```diff
@@ -156,8 +157,11 @@ void UpdateThumbColor()
 				}
 			}
+
+			Control.ThumbDrawable.SetColorFilter(Element.ThumbColor, FilterMode.SrcAtop);
```
why we are adding this ?

jsuarezruiz replied:
Removed.

rmarinho commented on src/Platform.Renderers/src/Xamarin.Forms.Platform.Android/AppCompat/SwitchRenderer.cs:
```diff
@@ -103,49 +104,49 @@ protected override void OnElementChanged(ElementChangedEventArgs<Switch> e)
 
 		protected override void OnElementPropertyChanged(object sender, PropertyChangedEventArgs e)
 		{
-			if (this.IsDisposed())
```
How are we sure we don't need this?

jsuarezruiz replied:
This line remains.

rmarinho reviewed (commented):
Seems we have some extra files, and i have some questions about changes

jsuarezruiz commented:
I have rebased and moved the macOS implementation to: https://github.com/xamarin/Xamarin.Forms/tree/switchhandler-mac in order to facilitate the revision and merge of this PR.

@rmarinho There should be no changes to any not needed files now.

jsuarezruiz commented:
Added _DO-NOT-MERGE!!!_ tag until adding handler tests.

mattleibow commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Core/ISwitch.cs:
```diff
@@ -0,0 +1,13 @@
+	public interface ISwitch : IView
+	{
+		bool IsToggled { get; set; }
+		Color OnColor { get; }
+		Color ThumbColor { get; }
```
Does there need to be a default/off color or is this just background color?

PureWeen replied:
Yea there's a conversation here about the OffColor

https://github.com/xamarin/Xamarin.Forms/pull/10355

I'm still a fan of just having a property called "TrackColor" and then we just get rid of OnColor

There's some conversation in the issue above around that.

We can still leave oncolor and offColor inside Forms but on the Maui side use TrackColor

mattleibow replied:
I see. Because of the wait the native things work - having colors fade and transition under the hood - it might make sense to keep those in the handler. There is talk of switching to brushes, but is it supported on any of the platforms? And also, is the VSM at the handlers level or just Controls? Not sure if Comet has a concept of VSM?

PureWeen replied:
As far as I remember Comet has something similar to VSM where you can define the state/color relationship

>  is the VSM at the handlers level or just Controls

Currently just at the Controls level. We had talked about this a little bit a couple months ago and the approach I've taken here is kind of the approach we were curious about. Could we just set the colors based on state and the handlers would then just "learn". So, the way I built this if the user is toggling on VSM then the ColorStateList will eventually match.  This does take one iteration to work but maybe we could eventually project the VSM over the handler on init? If we have a Controls.SwitchHandler that knows about VSM then that's a possible scenario.

I realize the comments on the Switch were saying to keep both because of native transitions but this same idea kind of extrapolates out across every single control right? Pretty much all the coloring APIs on iOS/UWP/Android do it via some sort of state based coloring. This isn't unique to switch

PureWeen replied:
@mattleibow even if we don't want to be smart about matching ColorStateLists to current states it's just nicer in general to use ColorStateLists for setting the color for all states. If I want red I just set a CSL once for all the state attributes and then it just works.

When using the color filter you have to call "SetColorFilter" every time the color changes
https://github.com/xamarin/Xamarin.Forms/pull/12196/files#diff-5f11a6818b5ece8c3e67df6b8e71a734267ac65f49c428525efb61ec8122a08bR74

To "re-apply" the color the user want and override the underlying CSL

PureWeen reviewed (changes requested):
I'm fixing the code that tests for colors

mattleibow reviewed (changes requested):
@PureWeen just pointed out that this should be `SwitchCompat`

PureWeen commented:
Should be fixed now @mattleibow

PureWeen commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/SwitchExtensions.cs:
```diff
@@ -0,0 +1,64 @@
+{
+	public static class SwitchExtensions
+	{
+		static bool ChangedThumbColor;
```
I don't think this can be here. On the renderer this was a per instance value so this can't be a static shared across all the handlers

I wonder for the ThumbDrawable if we can use `SetTintList` on the drawable and define the different colors for those states

For API 23+ I think we can use these APIs

https://developer.android.com/reference/android/widget/Switch#setThumbTintList(android.content.res.ColorStateList)

https://developer.android.com/reference/android/widget/Switch#setTrackTintList(android.content.res.ColorStateList)

PureWeen commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/SwitchExtensions.cs:
```diff
@@ -0,0 +1,254 @@
+			}
+		}
+
+		public static void UpdateOnColorSimple(this ASwitch aSwitch, ISwitch view)
```
@mattleibow @jsuarezruiz so this is the really simple one that works and just toggles based on state. So this is called every single time the checked gets changeed

mattleibow replied:
How does this work when the switch is halfway? Is it different to using the state lists?

PureWeen replied:
Not sure, I haven't stared deeply into the colors yet :-)  In a perfect world we could project the VSM over the native platform. 

Though, I think with either implementation we'll need some aspect of what I've done in this PR with extracting the colors out via the attributes and copying the android source.  I'm not really sure how else we can rewind the drawable tints to default. The only other approach I found to work was to make a deep copy of the drawable but that seems like a bad idea from a performance stand point

PureWeen commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/SwitchExtensions.cs:
```diff
@@ -0,0 +1,254 @@
+		}
+
+		//static Drawable? _defaultTrackDrawable;
+		public static void UpdateOnColor(this ASwitch aSwitch, ISwitch view)
```
@mattleibow @jsuarezruiz With this route we can just set the ColorStateList and then let it just do its thing.

We no longer have to call into UpdateOnColor every time the check changes

PureWeen commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/Android/SwitchExtensions.cs:
```diff
@@ -0,0 +1,254 @@
+	{
+		static ColorTrackingColorStateList? mSwitchTrackStateList;
+		// taken from android sourcec ccode
+		static ColorTrackingColorStateList getSwitchTrackColorStateList(this Context context)
```
@mattleibow @jsuarezruiz I think we'll need some version of this code regardless if we use ColorStateLists or not

Unless you all can find a better approach for setting the SwitchTrack Drawable back to `Color.Default`

jsuarezruiz commented on src/Platform.Handlers/src/Xamarin.Platform.Handlers/Platform/iOS/SwitchExtensions.cs:
```diff
@@ -0,0 +1,66 @@
+			if (NativeVersion.IsAtLeast(13))
+				uIView = uiSwitch.Subviews[0].Subviews[0];
+			else
+				uIView = uiSwitch.Subviews[0].Subviews[0].Subviews[0];
```
Could we create a` UIView FindChildren <T>` or similar extension method? It would be useful in other cases, and does this type of case where we need to access a specific child easier.

PureWeen replied:
This feels fairly edge case since I'm basically just accessing a specific child and I'm not "Searching" for a specific type

jsuarezruiz commented on src/Platform.Handlers/tests/Xamarin.Platform.Handlers.DeviceTests/AssertionExtensions.Android.cs:
```diff
@@ -45,12 +47,20 @@ public static AColor ColorAtPoint(this Bitmap bitmap, int x, int y, bool include
 		public static Bitmap ToBitmap(this AView view)
 		{
+			var layout = new FrameLayout(view.Context);
+			layout.LayoutParameters = new FrameLayout.LayoutParams(500, 500);
```
Why 500 as height and width?

PureWeen replied:
just randomly picked something 
No science to this really. 

It's just the container frame and then the added control will just calculates its own height/width

If we get to a point where we're screen shotting largers views we need to take pictures of we can feed these properties in

jsuarezruiz reviewed (commented):
Looks very nice, the handler test passes.
